stable/database/files/nuobackup

```diff
@@ -161,6 +149,86 @@ function execute_backup() {
    fi
 }
 
+function full_exist() {
+   local error
+
+   error="$( nuodocker get current-backup \
+      --db-name "${db_name}" \
+      --labels "${labels}" 2>&1)"
+   if [ $? -ne 0 ] && echo "$error" | grep -q "No existing backup path found"; then
+      # Check if full backup has been started already for all archiveIds. It
+      # won't be the case in the following scenarious:
+      # * initial database backup request
+      # * backup request after the database have been restored
+      # * new HC SM has been started in the backup group since the last full
+      #   backup
+      # * the backup group selection has been changed
+      return 1
+   fi
+   return 0
+}
+
+function last_full_failed() {
+   local backupset
+   local failed
+
+   backupset=$(get_current_backup)
+   failed=$(nuocmd get value --key "$NUODB_BACKUP_KEY/$db_name/${backup_group}/failed")
+   if [ -n "$backupset" ] && echo "$failed" | grep -q "${backupset}"; then
+      return 0
+   fi
+   return 1
+}
+
+function execute_full() {
+   execute_backup "full"
+}
+
+function execute_incremental() {
+   local backupset
+
+   if ! full_exist; then
+      echo "Executing full hotcopy as a prerequisite for incremental hotcopy"
+      execute_full
+   elif last_full_failed; then
+      # Incremental hotcopy requires a full hotcopy element in the backupset;
+      # in case the full hotcopy has failed, all incrementals using that
+      # backupset will fail until next full hotcopy is scheduled; if the
+      # failure is transient or user corrected the problem (like freeing some
+      # disk space), then executing another full will have a great chance to
+      # finish
+      backupset=$(get_current_backup)
+      echo "Executing full hotcopy as a prerequisite for incremental hotcopy: full hotcopy in backupset ${backupset} has failed"
+      execute_full
+   else
+      execute_backup "incremental"
+   fi
+}
+
+function execute_journal() {
+   local rc 
+
+   if ! full_exist; then
+      echo "Executing full hotcopy as a prerequisite for journal hotcopy"
+      execute_full
+   else 
+      execute_backup "journal"
+      rc=$?
+      if [ $rc -ne 0 ]; then
+         if echo "$backup_database_output" | grep -q "Full or incremental hot copy must be completed before journal hot copy"; then
+            # journal hot copy has been temporarly suspended due to archive sync;
+            # perform incremental hot copy first
+            echo "Executing incremental hot copy as journal hot copy is temporarily suspended"
+            execute_incremental || return $?
```

Comment:
Will journal hotcopy fail even if there is a full, because there are gaps in the journal (e.g. because all SMs were restarted)? If so, then it looks like you are relying on incremental to cover that gap before performing journal hotcopy. Is that correct?

Reply:
> Will journal hotcopy fail even if there is a full, because there are gaps in the journal (e.g. because all SMs were restarted)?

If the _full_ has been done after the restarted SM finished syncing, then we won't get into this _if_ block.

If some of the SMs in the group have been restarted and went to syncing, then the journal hot copy for all SMs in the group will be temporarily suspended. A _full_ or _incremental_ hot copy must be requested to form a new baseline for subsequent _journal_ hot hopy.